extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java

```diff
@@ -169,20 +172,32 @@ void processGeneratedBeans(CombinedIndexBuildItem index, BuildProducer<Annotatio
             if (!excluded) {
                 log.debugf("Registering generated gRPC bean %s that will delegate to %s", generatedBean, userDefinedBean);
                 delegatingBeans.produce(new DelegatingGrpcBeanBuildItem(generatedBean, userDefinedBean));
-                Set<String> blockingMethods = gatherBlockingMethodNames(userDefinedBean, index.getIndex());
-
+                Set<String> blockingMethods = gatherBlockingOrVirtualMethodNames(userDefinedBean, index.getIndex(), false);
+                Set<String> virtualMethods = gatherBlockingOrVirtualMethodNames(userDefinedBean, index.getIndex(), true);
                 generatedBeans.put(generatedBean.name(), blockingMethods);
+                virtuals.put(generatedBean.name(), virtualMethods);
             }
         }
 
-        if (!generatedBeans.isEmpty()) {
+        if (!generatedBeans.isEmpty() || !virtuals.isEmpty()) {
             // For every suitable bean we must:
             // (a) add @Singleton and @GrpcService
             // (b) register a BindableServiceBuildItem, incl. all blocking methods (derived from the user-defined impl)
-            for (Entry<DotName, Set<String>> entry : generatedBeans.entrySet()) {
-                BindableServiceBuildItem bindableService = new BindableServiceBuildItem(entry.getKey());
-                for (String blockingMethod : entry.getValue()) {
-                    bindableService.registerBlockingMethod(blockingMethod);
+            Set<DotName> names = new HashSet<>(generatedBeans.keySet());
+            names.addAll(virtuals.keySet());
+            for (DotName name : names) {
+                BindableServiceBuildItem bindableService = new BindableServiceBuildItem(name);
+                var blocking = generatedBeans.get(name);
+                var rovt = virtuals.get(name);
+                if (blocking != null) {
+                    for (String blockingMethod : blocking) {
+                        bindableService.registerBlockingMethod(blockingMethod);
+                    }
+                }
+                if (rovt != null) {
+                    for (String blockingMethod : rovt) {
```

Comment:
It's a nitpick ... but perhaps still call it `virtualMethod` ...

Reply:
fixed